src/main/java/org/zalando/nakadi/util/TimeLogger.java

```diff
@@ -0,0 +1,130 @@
+            if (currentStep >= this.stepStarts.length) {
+                this.stepStarts = Arrays.copyOf(this.stepStarts, this.stepStarts.length + 10);
+                this.stepNames = Arrays.copyOf(this.stepNames, this.stepStarts.length);
+            }
+            this.stepStarts[currentStep] = System.currentTimeMillis();
+            this.stepNames[currentStep] = message;
+            currentStep++;
+        }
+
+        String describe() {
+            final StringBuilder builder = new StringBuilder("[" + stepNames[0] + "]");
+            for (int i = 1; i < currentStep - 1; ++i) {
+                builder.append(" [")
+                        .append(stepNames[i])
+                        .append("=")
+                        .append(stepStarts[i + 1] - stepStarts[i])
+                        .append("]");
+            }
+            return builder.toString();
+        }
+    }
+
+    private static final ThreadLocal<LogCollector[]> THREAD_COLLECTORS = new ThreadLocal<>();
+    private static final LinkedList<LogCollector> COLLECTOR_POOL = new LinkedList<>();
+
+    public static void startMeasure(final String heading, final String stepName) {
+        LogCollector collector;
+        synchronized (COLLECTOR_POOL) {
+            collector = COLLECTOR_POOL.poll();
+        }
+        if (null == collector) {
+            collector = new LogCollector();
+        }
+        LogCollector[] collectors = THREAD_COLLECTORS.get();
+        final int idxToSet;
+        if (null == collectors) {
+            collectors = new LogCollector[2];
+            THREAD_COLLECTORS.set(collectors);
+            idxToSet = 0;
+        } else {
+            int freeSlot = -1;
+            for (int idx = 0; idx < collectors.length; ++idx) {
+                if (collectors[idx] == null) {
+                    freeSlot = idx;
+                    break;
+                }
+            }
+            if (freeSlot < 0) {
+                idxToSet = collectors.length;
+                collectors = Arrays.copyOf(collectors, collectors.length + 2);
+                THREAD_COLLECTORS.set(collectors);
+            } else {
+                idxToSet = freeSlot;
+            }
+        }
+        if (idxToSet > 0) {
+            collectors[idxToSet].addRecord(heading);
+        }
+        collectors[idxToSet] = collector;
+        collector.addRecord(heading);
+        collector.addRecord(stepName);
+    }
+
+    public static String finishMeasure() {
+        final LogCollector[] threadCollectors = THREAD_COLLECTORS.get();
+        if (null != threadCollectors) {
+            int idx = -1;
+            for (int i = 0; i < threadCollectors.length; ++i) {
+                if (null != threadCollectors[i]) {
+                    idx = i;
+                } else {
+                    break;
+                }
+            }
+            final LogCollector collector = idx >= 0 ? threadCollectors[idx] : null;
+            threadCollectors[idx] = null;
```

Comment:
@antban yeah, but result is different:

public class TimeLoggerTest {

    @Test
    public void test1() {
        TimeLogger.startMeasure("Outer", "doStuff1");
        // do stuff 1
        TimeLogger.addMeasure("stuff2");
        // do stuff 2
        TimeLogger.startMeasure("Inner", "stuff.2.1");
        // do stuff 2.1
        System.out.println(TimeLogger.finishMeasure());
        TimeLogger.addMeasure("stuff3");
        // do stuff 3
        System.out.println(TimeLogger.finishMeasure());
    }
}

Result:
`java.lang.NullPointerException
	at org.zalando.nakadi.util.TimeLogger.startMeasure(TimeLogger.java:79)
	at org.zalando.nakadi.util.TimeLoggerTest.test1(TimeLoggerTest.java:13)`